src/core/Ekubo.sol

```diff
@@ -0,0 +1,453 @@
+        // buyAmount = abi.decode(abi.decode(encodedBuyAmount, (bytes)), (uint256));
+        assembly ("memory-safe") {
+            // We can skip all the checks performed by `abi.decode` because we know that this is the
+            // verbatim result from `locked` and that `locked` encoded the buy amount
+            // correctly.
+            buyAmount := mload(add(0x60, encodedBuyAmount))
+        }
+    }
+
+    function _ekuboLockCallback(bytes calldata data) private returns (bytes memory) {
+        // We know that our calldata is well-formed. Therefore, the first slot is ekubo lock id,
+        // second slot is 0x20 and third is the length of the strict ABIEncoded payload
+        assembly ("memory-safe") {
+            data.length := calldataload(add(0x40, data.offset))
+            data.offset := add(0x60, data.offset)
+        }
+        return locked(data);
+    }
+
+    function _ekuboPay(
+        IERC20 sellToken,
+        address payer,
+        uint256 sellAmount,
+        ISignatureTransfer.PermitTransferFrom calldata permit,
+        bool isForwarded,
+        bytes calldata sig
+    ) private returns (uint256 payment) {
+        if (sellToken == ETH_ADDRESS) {
+            SafeTransferLib.safeTransferETH(payable(msg.sender), sellAmount);
+            return sellAmount;
+        } else {
+            // Encode the call plus the extra data that is going to be needed in the callback
+            bytes memory data;
+            assembly ("memory-safe") {
+                data := mload(0x40)
+
+                mstore(add(0x20, data), 0x0c11dedd) // selector for pay(address)
+                mstore(add(0x40, data), and(_ADDRESS_MASK, sellToken))
+
+                mstore(add(0x60, data), sellAmount)
+                let size := 0x44
+
+                // if permit is needed add it to data
+                if iszero(eq(payer, address())) {
+                    calldatacopy(add(0xa0, data), permit, 0x80)
+                    mstore(add(0x120, data), isForwarded)
+                    mstore(add(0x140, data), sig.length)
+                    calldatacopy(add(0x160, data), sig.offset, sig.length)
+                    size := add(size, add(0xc0, sig.length))
+                }
+
+                // update data length
+                mstore(data, size)
+
+                // update free memory pointer
+                mstore(0x40, add(data, add(0x20, size)))
+            }
+            bytes memory encodedPayedAmount =
+                _setOperatorAndCall(msg.sender, data, uint32(IEkuboCallbacks.payCallback.selector), payCallback);
+            assembly ("memory-safe") {
+                // We can skip all the checks performed by `abi.decode` because we know that this is the
+                // verbatim result from `payCallback` and that `payCallback` encoded the payment
+                // correctly.
+                payment := mload(add(0x60, encodedPayedAmount))
```

Comment:
Ditto other comment about verifying whether there are dirty bits or skipping that if we know dirty bits are impossible

Reply:
the comment is there already, ecodedPayment came from payCallback which is our function and encodes data properly